src/containers/LoginRegister/index.jsx

```diff
@@ -0,0 +1,224 @@
+import React, { useContext, useState } from "react";
+import Login from "../../components/Login";
+import Register from "../../components/Register";
+import PasswordReset from "../../components/PasswordReset";
+import firestore, { auth } from "../../firebaseConfig";
+import firebase from "firebase";
+import { AuthContext } from "../../auth/authContext";
+import { withRouter } from "react-router-dom";
+import { Alert, Spin } from "antd";
+import "./index.css";
+import { useTranslation } from "react-i18next";
+import kids from "../../images/LoginKids.svg";
+
+const LoginRegisterPage = (props) => {
+  const [t] = useTranslation();
+  const login = async (email, password) => {
+    setError("");
+    setMessage(<></>);
+    try {
+      setLoading(true);
+      await auth.signInWithEmailAndPassword(email, password);
+      setLoading(false);
+      props.history.push("/");
+    } catch (e) {
+      setLoading(false);
+      setError(e.message);
+    }
+  };
+  const passwordReset = async (email) => {
+    setError("");
+    setMessage(<></>);
+    try {
+      setLoading(true);
+      await auth.sendPasswordResetEmail(email);
+      setLoading(false);
+      setMessage(
+        <Alert
+          style={{ marginBottom: 10 }}
+          showIcon
+          message={
+            <div>
+              {t("loginRegister.anEmailsent")}
+              <b>{email}</b>
+              {t("loginRegister.anEmailsentcontain")}
+            </div>
+          }
+          type="success"
+        ></Alert>
+      );
+      setActive("login");
+    } catch (e) {
+      setLoading(false);
+      setError(e.message);
+    }
+  };
+  // Login with FACEBOOK
+  const loginWithFacebook = async () => {
+    setError("");
+    setMessage(<></>);
+    try {
+      setLoading(true);
+      let provider = new firebase.auth.FacebookAuthProvider();
+      provider.setCustomParameters({
+        display: "popup",
+      });
+      let u = await auth.signInWithPopup(provider);
+      console.log("facebook user: ", u.user);
+      setLoading(false);
+      props.history.push("/");
+    } catch (e) {
+      setLoading(false);
+      setError(e.message);
+    }
+  };
+  // Loogin with GOOGLE
+  const loginWithGoogle = async () => {
+    setError("");
+    setMessage(<></>);
+    try {
+      setLoading(true);
+      let provider = new firebase.auth.GoogleAuthProvider();
+      provider.addScope("https://www.googleapis.com/auth/contacts.readonly");
+      let u = await auth.signInWithPopup(provider);
+      console.log("google user: ", u.user);
+      setLoading(false);
+      props.history.push("/");
+    } catch (e) {
+      setLoading(false);
+      setError(e.message);
+    }
+  };
+
+  const register = async (fullname, email, age, password) => {
+    console.log(email, password);
+    setError("");
+    setMessage(<></>);
```

Comment:
see how these two lines are copy pasted everywhere

the rule is that if I wanted, for example, to change the empty message to be null instead of <>, I should only have to change it one place

Reply:
Hi @louisrli , I thought by using this I'll stop any error message before submitting the form, I'm a bit confused about how to change it as your comment, would you kindly guide me about it ?

Reply:
```
    setError("");
    setMessage(<></>);
    try {
      setLoading(true);
      await auth.signInWithEmailAndPassword(email, password);
      setLoading(false);
      props.history.push("/");
    } catch (e) {
      setLoading(false);
      setError(e.message);
    }
  };
```

look at these lines

the code itself is fine, the problem is that its duplicated and copy-pasted everywhere then one line changed, which is not okay.

suppose I changed your state variable name: setMessage to setErrorMessage (i just change the variable name, thats it)

how many places in the code do i have to go and change? the answer *should* be 1, but its like 5 right now, because you copy-pasted the same sequence of code everywhere

work with @tmansuroglu , he can explain it to you in-depth